File: src/main/java/com/example/estantevirtual/service/BookService.java
package com.example.estantevirtual.service;

import com.example.estantevirtual.exception.ResourceAlreadyExistsException;
import com.example.estantevirtual.model.Book;
import com.example.estantevirtual.model.Options;
import com.example.estantevirtual.repository.BookRepository;
import com.example.estantevirtual.utils.RedisCacheEvictor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
public class BookService {

    @Autowired
    BookRepository bookRepository;

    @Autowired
    RedisCacheEvictor redisCacheEvictor;

    @Cacheable(value = "books")
    public Optional<Book> findBookBy(UUID id) {
        return bookRepository.findById(id);
    }

    public void saveBook(Book book) {
        if (bookRepository.findByTitleAndAuthor(book.getTitle(), book.getAuthor()).isPresent()) {
            throw new ResourceAlreadyExistsException("Livro já cadastrado");
        }
        redisCacheEvictor.evictAllCaches();
        book.setId(UUID.randomUUID());
        bookRepository.save(book);
    }

    @Cacheable(value = "books")
    public Page<Book> findBooks(Options options) {
        validateParams(options);
        return bookRepository.findAll(PageRequest.of(
                options.getPage(),
                options.getPageSize(),
                Sort.Direction.valueOf(options.getSort()),
                options.getOrderBy()
        ));
    }

    public boolean deleteBook(UUID id) {
        if (bookRepository.findById(id).isPresent()) {
            redisCacheEvictor.evictAllCaches();
            bookRepository.deleteById(id);
            return true;
        }
        return false;
    }

    public Boolean updateBook(UUID id, Book book) {
        Optional<Book> bookToUpdate = bookRepository.findById(id);
        if (bookToUpdate.isPresent()) {
            bookToUpdate.get().setTitle(book.getTitle());
            bookToUpdate.get().setAuthor(book.getAuthor());
            redisCacheEvictor.evictAllCaches();
            bookRepository.save(bookToUpdate.get());
            return true;
        }
        return false;
    }

    public Map<String, Object> responseBuilder(Page<Book> book) {
        Map<String, Object> books = new LinkedHashMap<>();
        books.put("books", book.getContent());
        books.put("totalItems", book.getTotalElements());
        books.put("currentPage", book.getNumber());
        books.put("totalPages", book.getTotalPages());

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("data", books);
        return data;
    }

    private void validateParams(Options options) {
        if (options.getPage() < 0) {
            throw new IllegalArgumentException("O parâmetro page deve ser maior ou igual a zero");
        }
        if (options.getPageSize() < 0) {
            throw new IllegalArgumentException("O parâmetro pageSize deve ser maior ou igual a zero");
        }
        if (!options.getSort().equals(Sort.Direction.ASC.toString()) && !options.getSort().equals(Sort.Direction.DESC.toString())) {
            throw new IllegalArgumentException("O parâmetro sort deve ser ASC ou DESC");
        }
        if (!options.getOrderBy().equals("id") && !options.getOrderBy().equals("title") && !options.getOrderBy().equals("author")) {
            throw new IllegalArgumentException("O parâmetro orderBy deve ser id, title ou author");
        }
    }
}

File: src/main/kotlin/com/estantevirtual/service/BookService.kt
package com.estantevirtual.service

import com.estantevirtual.exception.IsbnException
import com.estantevirtual.exception.ResourceAlreadyExistsException
import com.estantevirtual.model.Book
import com.estantevirtual.model.Books
import com.estantevirtual.model.Options
import com.estantevirtual.repository.BookRepository
import com.estantevirtual.utils.RedisCacheEvict
import com.estantevirtual.validator.IsbnValidator
import com.estantevirtual.validator.OptionsValidator
import org.slf4j.Logger
import org.springframework.beans.BeanUtils
import org.springframework.cache.annotation.Cacheable
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.web.util.UriComponentsBuilder
import java.util.*

@Service
class BookService(
    private val bookRepository: BookRepository,
    private val redisCacheEvict: RedisCacheEvict,
    private val logger: Logger,
    private val isbnValidator: IsbnValidator,
    private val optionsValidator: OptionsValidator
) {
    @Cacheable(value = ["books"])
    fun findBookBy(isbn: String?): Optional<Book?>? {
        return bookRepository.findByIsbn(isbn)
    }

    @Cacheable(value = ["books"])
    fun findBooks(options: Options): Books {
        optionsValidator.validateParams(options)
        val books = bookRepository.findAll(
            PageRequest.of(
                options.page,
                options.pageSize,
                Sort.Direction.valueOf(options.sort!!),
                options.orderBy
            )
        )

        val nextPage = if (books.hasNext()) {
            buildPageUrl(options.page + 1, UriComponentsBuilder.fromPath("/v1/books"), options)
        } else null

        val prevPage = if (books.hasPrevious() && books.hasContent()) {
            buildPageUrl(options.page - 1, UriComponentsBuilder.fromPath("/v1/books"), options)
        } else null

        return Books(
            books,
            books.totalElements,
            options.page,
            options.pageSize,
            nextPage,
            prevPage
        )
    }

    fun save(book: Book?) {
        book ?: throw IllegalArgumentException("Book cannot be null")
        if (isbnValidator.isValid(book.isbn)) {
            val response = bookRepository.findByIsbn(book.isbn)
            if (response.isPresent) {
                throw ResourceAlreadyExistsException("Resource already exists")
            }
            bookRepository.save(book)
            logger.info("Created ISBN: ${book.isbn} ")
            redisCacheEvict.evictAllCaches()
        } else {
            throw IsbnException()
        }
    }

    fun update(isbn: String?, book: Book?): Boolean {
        book ?: throw IllegalArgumentException("Book cannot be null")
        val existingBook = bookRepository.findByIsbn(isbn)
        if (existingBook.isPresent) {
            val bookToUpdate = existingBook.get()
            BeanUtils.copyProperties(book, bookToUpdate, "isbn")
            bookRepository.save(bookToUpdate)
            logger.info("Updated ISBN: $isbn")
            redisCacheEvict.evictAllCaches()
            return true
        }
        return false
    }

    fun delete(isbn: String?): Boolean {
        if (bookRepository.findByIsbn(isbn).isPresent) {
            bookRepository.deleteByIsbn(isbn)
            logger.info("Deleted ISBN: $isbn")
            redisCacheEvict.evictAllCaches()
            return true
        }
        return false
    }

    private fun buildPageUrl(page: Int, uriBuilder: UriComponentsBuilder, options: Options): String {
        return uriBuilder.replaceQueryParam("page", page)
            .queryParam("per_page", options.pageSize)
            .queryParam("order_by", options.orderBy)
            .queryParam("sort", options.sort)
            .toUriString()
    }
}